File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <bits/stdc++.h>
using namespace std;

int dy[4] = {-1,0,1,0};
int dx[4] = {0,1,0,-1};

int arr[105][105];
int visited[105][105];
int a[105][105];
int N;
int cnt;

void dfs(int y, int x){
    visited[y][x] = 1;
    
    int ny;
    int nx;
    
    for(int i =0; i < 4; i++){
        ny = y + dy[i];
        nx = x + dx[i];
        if(nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
        if(visited[ny][nx] == 0 && a[ny][nx] == 1){
            dfs(ny,nx);
        }
    }
}


int main()
{
    
    cin >> N;
    
    int Max = 0;
    int Min = 100;
    
    for(int i = 0; i < N; i++){
        for(int j = 0; j < N; j++){
            cin >> arr[i][j];
            Min = min(Min,arr[i][j]);
            Max = max(Max,arr[i][j]);
        }
    }
    
    
    int FMax = 1;
    
    for(int H = Min; H <= Max; H++){
        cnt = 0;
        fill(&a[0][0], &a[0][0] + 105*105, 0);
        fill(&visited[0][0], &visited[0][0] + 105*105, 0);
        
        
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                if(arr[i][j] > H){
                    a[i][j] = 1; 
                }
            }
        }
        
        
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                if(a[i][j] == 1 && visited[i][j] == 0){
                    dfs(i,j);
                    cnt++;
                }
            }
        }
        
        
        FMax = max(FMax, cnt);
        
    }
    
    cout << FMax;
    
    return 0;
}

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.util.*;
import java.io.*;


public class Main {
	
	static int N;
	static int[][] arr;
	static boolean[][] visited;
	static int ans;
	static int[] dcol = {0,0,-1,1};  // 상  하  좌   우
	static int[] drow = {-1,1,0,0};
	static int Min = 1;
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		N = Integer.parseInt(br.readLine());
		
		arr = new int[N][N];
		
        int minHeight = 101;
        int maxHeight = 0;
        
        
		for(int i=  0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for(int j = 0; j < N; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
				minHeight = Math.min(minHeight, arr[i][j]);
                maxHeight = Math.max(maxHeight, arr[i][j]);
			}
		}
		
		
		//System.out.println(list.size());
		
		
		
		for(int num = minHeight; num <= maxHeight; num++) {
			
			visited = new boolean[N][N];
			
			
			
			ans = 0;
			for(int i = 0; i < N; i++) {
				for(int j = 0; j < N; j++) {
					if(chk(i,j) && arr[i][j] > num) {
						go(i,j,num);
						ans++;
					}

				}
			}
			Min = Math.max(Min, ans);
		}
		
		
		System.out.println(Min);
		
	} // main
	
	
	
	
	static void go(int row, int col,int h) {
		
		visited[row][col] = true;
		
		for(int i = 0; i < 4; i++) {
			int nextrow = row + drow[i];
			int nextcol = col + dcol[i];
			
			if(chk(nextrow,nextcol) && arr[nextrow][nextcol] > h) {
				go(nextrow, nextcol,h);
			}
			
		}
		
	} // go
	
	
	
	static boolean chk(int row, int col) {
		return row >= 0 && row < N && col >= 0 && col < N && !visited[row][col];
	} // chk
	
	
}